### PR TITLE
Extends fetch_asset core method to find assets and cancel the test. [v4]

### PR DESCRIPTION
**Pull request by willianrampazzo**
This extension allows the user to choose if the `fetch_asset` core method should move the asset to the cache or should just look for the asset in the cache.
    
The user can cancel the test when the asset is not available in the cache or when there was a problem adding the asset to the cache.

Signed-off-by: Willian Rampazzo <willianr@redhat.com>

---
Changes from v3:
- Changed the behavior of `fetch_asset_file` on Asset class to raise an exception instead of returning `None`.
- Handle the exception from `fetch_asset_file` on `fetch_asset` core method.
- Refactor `get_metadata` to reduce code.
- Add comments.

Changes from v2:
- Remove the explicit calls to the cancel exception.
- Fix string formating for the log message.
- Add functional tests.

Changes from v1:
- Make it possible for the user to cancel the test when the asset was not available in the cache or when there was a problem adding the asset to the cache.

### CODE REVIEW

**Comment by willianrampazzo** [comment]
Relates to #3600.

File: avocado/utils/asset.py [hunk 1]
```diff
@@ -361,13 +367,15 @@ def get_metadata(self):
...
-            if os.path.isfile(metadata_file):
-                with open(metadata_file, "r") as f:
-                    metadata = json.load(f)
-                    return metadata
-        return None
+        try:
+            asset_file = self.find_asset_file()
+        except OSError:
+            LOG.info("Metadata not available.")
+            return None
```
Comment (willianrampazzo) [inline avocado/utils/asset.py]: There is a possible improvement here to change the behavior of `get_metadata` and throw an exception when the metadata is not available to follow the pattern from the class, but this is not related to the objective of this PR, so I'm taking note to change it as soon as #3600 is closed

File: avocado/core/test.py [hunk 2]
```diff
@@ -1093,13 +1094,51 @@ def fetch_asset(self, name, asset_hash=None, algorithm=None,
...
+            # return the path to the asset when fetch succeeded
+            asset_path = asset_obj.fetch()
+            return asset_path
+        except OSError as e:
+            # if there was a problem fetching the asset
+            if cancel_on_missing:
+                # cancel when requested
+                self.cancel(missing_asset_message)
+            # otherwise re-throw OSError
+            raise e
```
Comment (apahim) [inline avocado/core/test.py]: Nice, thanks! Now users will have no surprises when using the new feature :)
Do you plan to de-duplicate the duplicated block?

Comment (willianrampazzo) [reply avocado/core/test.py]: Working on it right now, stay tuned for the next minutes ;)

**Comment by willianrampazzo** [comment]
Closed in favor of #3622.
